### DISCUSSION TITLE
Wrong parsing requestBody and Where query.object

### DISCUSSION THREAD

**Discussion by peace195** [loopbackio/loopback-next#7224] [Q&A] [answered]
Here is the example code: https://github.com/strongloop/loopback4-example-shopping/blob/master/packages/shopping/src/controllers/user-order.controller.ts#L104

How to reproduce the issue:
1. Add `@inject(SecurityBindings.USER) currentUserProfile: UserProfile`  
2. Call `PATCH /users​/{userId}​/orders` 

Attachment is the code, its swagger and console.
I passed `{"id": 1}` as Where object. `{"total": 100}` as Order object of requestBody().
But in backend, it parsed `{ id: 1 }` as Order object and `{ total: 100 }` as Where filter.

**I faced this issue in both loopback4-example-shopping project and my other project**

<img width="908" alt="Screen Shot 2021-03-13 at 13 20 31" src="https://user-images.githubusercontent.com/8999858/111021811-92f07d00-8401-11eb-9777-65f2c7e2632c.png">
<img width="1298" alt="Screen Shot 2021-03-13 at 13 20 57" src="https://user-images.githubusercontent.com/8999858/111021814-96840400-8401-11eb-9e61-5c4739104efc.png">
<img width="692" alt="Screen Shot 2021-03-13 at 13 21 15" src="https://user-images.githubusercontent.com/8999858/111021817-97b53100-8401-11eb-98bf-c616ce3d91a1.png">

**Comment by raymondfeng** [comment]
This is pretty weird. What curl command does the Swagger UI generate? I can investigate if you can create an simple app with just that route and share the git repo.

**Reply by peace195** [reply]
@raymondfeng Thank you very much.
Yes, it is:
```
http://localhost:3000/users/604dadaf126de4845d0eef3/orders?where={
  "id": 1
}
```

Here is my example repo: https://github.com/peace195/loopback-PATCH-example
The step to reproduce:
1. `npm i`
2. `npm run docker:start`
3. `npm start`
4. Go to http://localhost:3000/shoppy.html to register an account.
5. Go to http://localhost:3000/explorer/#/UserManagementController/UserManagementController.login to login and get token
6. Go to http://localhost:3000/explorer/#/UserManagementController/UserManagementController.printCurrentUser , authen with token, then get `userId`.
7. Go to http://localhost:3000/explorer/#/UserOrderController/UserOrderController.patchOrders to execute the PATCH and get the issue.

<img width="817" alt="Screen Shot 2021-03-14 at 08 50 42" src="https://user-images.githubusercontent.com/8999858/111054848-7eb08c80-84a2-11eb-80d9-ec2a8603f116.png">

<img width="1432" alt="Screen Shot 2021-03-14 at 13 31 39" src="https://user-images.githubusercontent.com/8999858/111059764-a582b900-84ca-11eb-84d4-b2bec6a5075b.png">
<img width="1140" alt="Screen Shot 2021-03-14 at 13 31 51" src="https://user-images.githubusercontent.com/8999858/111059768-ad425d80-84ca-11eb-8c5a-bfeeb48dd208.png">
<img width="1155" alt="Screen Shot 2021-03-14 at 13 32 05" src="https://user-images.githubusercontent.com/8999858/111059769-af0c2100-84ca-11eb-8fb2-8e48989f211c.png">
<img width="1146" alt="Screen Shot 2021-03-14 at 13 32 18" src="https://user-images.githubusercontent.com/8999858/111059771-b03d4e00-84ca-11eb-8941-cd6d2f750281.png">
<img width="1148" alt="Screen Shot 2021-03-14 at 13 32 44" src="https://user-images.githubusercontent.com/8999858/111059772-b16e7b00-84ca-11eb-8b4d-51db7e266993.png">

**Reply by peace195** [reply]
@raymondfeng FYI: when remove this line: https://github.com/peace195/loopback-PATCH-example/blob/master/src/controllers/user-order.controller.ts#L106 , everything will be good. But why adding this line give us wrong parsing? 

Thank you very much 🙏

**Comment by raymondfeng** [comment, marked as answer]
I can reproduce the problem and confirm it's a bug in REST parameter parsing. A quick workaround is to move `currentUserProfile` to be the last argument.

**Reply by raymondfeng** [reply]
See a fix at https://github.com/strongloop/loopback-next/pull/7251

**Reply by peace195** [reply]
@raymondfeng  Thank you very much ✋
